Firestore/core/src/firebase/firestore/util/vector_of_ptr.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_VECTOR_OF_PTR_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_UTIL_VECTOR_OF_PTR_H_
+
+#include <initializer_list>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "absl/algorithm/container.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+/**
+ * A std::vector of some pointer type where equality and many other operations
+ * are defined as operating on the value pointed to rather than on the pointers
+ * themselves.
+ *
+ * Contrast with `std::vector<std::shared_ptr<T>>`, where `operator==` just
+ * checks if the pointers in the collection are equal rather than checking if
+ * the things the pointers point to are equal.
+ *
+ * This is useful in cases where values of type T need to be held by pointer
+ * for some reason, usually this is to enable polymorphism or because copying
+ * values of T is expensive.
+ */
+template <typename P>
+class vector_of_ptr {
+ public:
+  using pointer_type = P;
+  using value_type = decltype(*std::declval<P>());
+  using vector_type = std::vector<P>;
+
+  vector_of_ptr() = default;
+  vector_of_ptr(std::initializer_list<P> values) : values_(std::move(values)) {
+  }
+
+  size_t size() const {
+    return values_.size();
+  }
+
+  void push_back(P value) {
+    values_.push_back(std::move(value));
+  }
+
+  typename vector_type::iterator begin() {
+    return values_.begin();
+  }
+  typename vector_type::const_iterator begin() const {
+    return values_.begin();
+  }
+
+  typename vector_type::iterator end() {
+    return values_.end();
+  }
+  typename vector_type::const_iterator end() const {
+    return values_.end();
+  }
+
+  friend bool operator==(const vector_of_ptr& lhs, const vector_of_ptr& rhs) {
```

Comment:
Question: it seems like this won't work correctly on polymorphic types? Am I missing something? Otherwise, would doing a check like `static_assert(!std::is_polymorphic<vector_of_ptr::value_type>::value, "operator== will not work correctly on polymorphic values")` be a good idea?

Reply:
`operator==` does not work generally for anything polymorphic -- which version to call is resolved statically by the compiler.

Polymorphic equality has to be handled by delegation to a virtual function in any type hierarchy that supports it. So, for example, the way this works in `FieldValue` is that there's a single definition of `operator==`:

```c++
bool operator==(const FieldValue& lhs, const FieldValue& rhs) {
  return lhs.rep_->Equals(*rhs.rep_);
}
```

And `FieldValue::BaseValue` (the `rep_` in the above), defines `Equals` like so:

```c++
    virtual Type type() const = 0;

    virtual bool Equals(const BaseValue& other) const = 0;
```

The implementation of `Equals` in any subclass has to do a type check before comparing values:

```c++
class DoubleValue : public NumberValue<Type::Double, double> {
  // ...
  bool Equals(const BaseValue& other) const override {
    if (type() != other.type()) return false;

    auto& other_value = Cast<DoubleValue>(other);
    return util::DoubleBitwiseEquals(value(), other_value.value());
  }
};
```

This is slightly different than the way we'll handle `Field` or `Document` types where the `std::shared_ptr` will be public. For example for `vector_of_ptr<std::shared_ptr<Field>>`, `*left == *right` will resolve staticically to `operator==(const Field&, const Field&)`. That will dispatch similarly to what we have in `FieldValue`.

The next PR that ports `FSTFilter` to C++ `Filter` implements this.

Reply:
As long as this is taken into consideration, I'm fine with it. Feel free to decline on my suggestion re. the `static_assert` if you think it's unnecessarily restrictive.

Reply:
I can't do the `static_assert` you're suggesting because I'm writing this to support collections of `Filter` instances, which are held by pointer only because they're polymorphic :-).